maitake/src/wait/map.rs

```diff
@@ -0,0 +1,926 @@
+    Closed,
+    AlreadyConsumed,
+    NeverAdded,
+}
+
+/// An indication of the result of a call to `wait()`
+// Note: We duplicate the WaitResult type as it currently only has "Closed"
+// as an error condition. We should merge these two on the next breaking
+// change.
+pub type WaitResult<T> = Result<T, WaitError>;
+
+const fn closed<T>() -> Poll<WaitResult<T>> {
+    Poll::Ready(Err(WaitError::Closed))
+}
+
+const fn consumed<T>() -> Poll<WaitResult<T>> {
+    Poll::Ready(Err(WaitError::AlreadyConsumed))
+}
+
+const fn never_added<T>() -> Poll<WaitResult<T>> {
+    Poll::Ready(Err(WaitError::NeverAdded))
+}
+
+const fn notified<T>(data: T) -> Poll<WaitResult<T>> {
+    Poll::Ready(Ok(data))
+}
+
+/// A map of [`Waker`]s implemented using an [intrusive doubly-linked
+/// list][ilist].
+///
+/// A `WaitMap` allows any number of tasks to [wait] asynchronously and be
+/// woken when a value with a certain key arrives. This can be used to
+/// implement structures like "async mailboxes", where an async function
+/// requests some data (such as a response) associated with a certain
+/// key (such as a message ID). When the data is received, the key can
+/// be used to provide the task with the desired data, as well as wake
+/// the task for further processing.
+///
+/// # Examples
+///
+/// Waking a single task at a time by calling [`wake`][wake]:
+///
+/// ```
+/// use std::sync::Arc;
+/// use maitake::{scheduler::Scheduler, wait::map::{WaitMap, WakeOutcome}};
+///
+/// const TASKS: usize = 10;
+///
+/// // In order to spawn tasks, we need a `Scheduler` instance.
+/// let scheduler = Scheduler::new();
+///
+/// // Construct a new `WaitMap`.
+/// let q = Arc::new(WaitMap::new());
+///
+/// // Spawn some tasks that will wait on the queue.
+/// // We'll use the task index (0..10) as the key.
+/// for i in 0..TASKS {
+///     let q = q.clone();
+///     scheduler.spawn(async move {
+///         let val = q.wait(i).await.unwrap();
+///         assert_eq!(val, i + 100);
+///     });
+/// }
+///
+/// // Tick the scheduler once.
+/// let tick = scheduler.tick();
+///
+/// // No tasks should complete on this tick, as they are all waiting
+/// // to be woken by the queue.
+/// assert_eq!(tick.completed, 0, "no tasks have been woken");
+///
+/// // We now wake each of the tasks, using the same key (0..10),
+/// // and provide them with a value that is their `key + 100`,
+/// // e.g. 100..110. Only the task that has been woken will be
+/// // notified.
+/// for i in 0..TASKS {
+///     let result = q.wake(&i, i + 100);
+///     assert!(matches!(result, WakeOutcome::Woke));
+///
+///     // Tick the scheduler.
+///     let tick = scheduler.tick();
+///
+///     // Exactly one task should have completed
+///     assert_eq!(tick.completed, 1);
+/// }
+///
+/// // Tick the scheduler.
+/// let tick = scheduler.tick();
+///
+/// // No additional tasks should be completed
+/// assert_eq!(tick.completed, 0);
+/// assert!(!tick.has_remaining);
+/// ```
```

Comment:
nitpicky, not a big deal: is there a similar test in the regular `tests` module? i think doctests won't get run with miri etc.

Reply:
Yup! This is basically a duplicate of `maitake::wait::map::tests::wake_one()`.